apps/hash-graph/.justfile

```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env just --justfile
+
+set fallback
+set dotenv-load
+
+repo := `git rev-parse --show-toplevel`
+profile := env_var_or_default('PROFILE', "dev")
+
+export HASH_GRAPH_PG_DATABASE := env_var('HASH_GRAPH_PG_DEV_DATABASE')
+export DOCKER_BUILDKIT := "1"
+
+[private]
+@default:
+  just usage
+
+# Runs the Graph API and accompanying services
+run *arguments:
+  cargo run --profile {{profile}} --bin hash-graph -- {{arguments}}
+
+# Spins up the deployment environment
+deployment-up *arguments:
+  @just yarn external-services up postgres type-fetcher --build --wait
+  @just yarn external-services up graph-migrate {{arguments}}
+
+# Tears down the deployment environment
+deployment-down *arguments:
+  @just yarn external-services down {{arguments}}
+
+# Generates the OpenAPI client for the Graph REST API
+generate-openapi-client:
+  @just deployment-up graph --wait
+  just yarn workspace @local/hash-graph-client-generator generate
+  just yarn workspace @local/hash-graph-client prettier --write .
+  just yarn workspace @local/hash-graph-client fix:eslint
+  @just deployment-down
+
+[private]
+test *arguments:
+  @just deployment-up
+  @just --justfile {{repo}}/.justfile test {{arguments}}
+  @just deployment-up graph --wait
+  just yarn httpyac send --all {{repo}}/apps/hash-graph/tests/rest-test.http
+  just generate-openapi-client
+  @just deployment-down
+
+[private]
+coverage *arguments:
+  @just deployment-up
+  cargo llvm-cov --workspace --all-features --all-targets {{arguments}}
+  @just deployment-down
+
+[private]
+bench *arguments:
+  @just deployment-up
+  @just --justfile {{repo}}/.justfile bench {{arguments}}
```

Comment:
Any possible alternatives to this?

Not really a fan of the implicit complexity we're adding in about reusing recipes, but I'm not sure we can easily get around it.

It seems that sometimes things reuse by just's parent resolution logic, and sometimes by explicitly calling one..

Reply:
Actually, I like this solution. It's basically extending the default behavior. In comparison to `cargo make`, where we had two tasks for each step to make overwriting slightly easier, it was still hard to get the behavior as it only was partially overwriting the job.

One solution would be to have a general job, which would be run prior to and after every job, which can be overwritten, something like `pre-bench` and `post-bench`. Probably a solution we can discover but most likely not worth doing it right now, especially as we have to discover how we will invoke `turbo` in CI.

Reply:
I'm not arguing that the `cargo-make` approach was a good one. Just that it's proving quite hard to actually discover what is ran if I run a `just` command. I have to trace it through all of the parent `.justfiles` and then deal with logic around `[no-cd]` on _some_ of them, and now I **also** need to look at how some of them skips that whole process to directly call a specific `justfile`, which is going to create another set of edge-cases we need to handle when adjusting top-level definitions.

This is creating a very opaque and potentially convoluted dependency graph.

Reply:
The job is called the same, so just invoking `just bench` would be recursion. I don't know how we could re-use that behavior. Things like
```just
@`(cd .. && just bench {{arguments}})`
```
does not work either as the working directory would be different. One option is to implement a recipe, which does this recursion logic until it finds a parent `.justfile`, but this I consider out of scope of this PR.

Reply:
Yep, I can't think of an easy way either. I think the `pre-X` and `post-X` pattern might be a good idea to explore later as well, but it's probably got a horrible amount of boilerplate